README.md
# CipherPayload

[![made-with-Go](https://img.shields.io/badge/Made%20with-Go-1f425f.svg)](http://golang.org) [![Go Reference](https://pkg.go.dev/badge/github.com/buildingwatsize/cipherPayload@v0.1.0.svg)](https://pkg.go.dev/github.com/buildingwatsize/cipherPayload@v0.1.0) [![GitHub issues](https://img.shields.io/github/issues/buildingwatsize/cipherPayload)](https://github.com/buildingwatsize/cipherPayload/issues) [![GitHub forks](https://img.shields.io/github/forks/buildingwatsize/cipherPayload)](https://github.com/buildingwatsize/cipherPayload/network) [![GitHub stars](https://img.shields.io/github/stars/buildingwatsize/cipherPayload)](https://github.com/buildingwatsize/cipherPayload/stargazers)

CipherPayload middleware for Fiber that use AES Algorithm for encrypt and decrypt payload in request and response body.

## Table of Contents

- [CipherPayload](#cipherpayload)
  - [Table of Contents](#table-of-contents)
  - [Installation](#installation)
  - [Signatures](#signatures)
  - [Examples](#examples)
  - [Config](#config)
  - [Default Config](#default-config)
  - [Default Response](#default-response)
  - [KeyPairs Property](#keypairs-property)
  - [Payload Template](#payload-template)
    - [Request](#request)
    - [Response](#response)
  - [Example Usage](#example-usage)

## Installation

```bash
  go get -u github.com/buildingwatsize/cipherPayload
```

## Signatures

```go
func New(config ...Config) fiber.Handler
```

## Examples

Import the middleware package that is part of the Fiber web framework

```go
import (
  "github.com/gofiber/fiber/v2"
  "github.com/buildingwatsize/cipherPayload"
)
```

After you initiate your Fiber app, you can use the following possibilities:

```go
// Default middleware config
app.Use(cipherPayload.New(cipherPayload.Config{
  KeyPairs: cipherPayload.KeyPairs{
    AESKeyForEncrypt: []byte("AES_KEY_FOR_ENCRYPT"),
    AESIVForEncrypt:  []byte("AES_IV_FOR_ENCRYPT"),
    AESKeyForDecrypt: []byte("AES_KEY_FOR_DECRYPT"),
    AESIVForDecrypt:  []byte("AES_IV_FOR_DECRYPT"),
  },
}))

// Or extend your config for customization
app.Use(cipherPayload.New(cipherPayload.Config{
  KeyPairs: cipherPayload.KeyPairs{
    AESKeyForEncrypt: []byte("AES_KEY_FOR_ENCRYPT"),
    AESIVForEncrypt:  []byte("AES_IV_FOR_ENCRYPT"),
    AESKeyForDecrypt: []byte("AES_KEY_FOR_DECRYPT"),
    AESIVForDecrypt:  []byte("AES_IV_FOR_DECRYPT"),
  },
  AllowMethod: []string{"POST", "OPTIONS"},
  DebugMode: true,
}))
```

## Config

```go
// Config defines the config for middleware.
type Config struct {
  // Next defines a function to skip this middleware when returned true.

  // Optional. Default: nil
  Next func(c *fiber.Ctx) bool

  // Required. Default: KeyPairs{}
  KeyPairs KeyPairs

  // Optional. Default: ["OPTIONS", "POST", "PUT", "DELETE"]
  AllowMethod []string

  // Optional. Default: false
  DebugMode bool

  // Optional. [Default: false]
  StrictMode bool

  // Optional. Default: true
  ExcludeHealthAPI bool

  // Optional. Default: BadRequestResponse
  FailResponse func(c *fiber.Ctx, msg string) error

  // Optional. Default: InternalServerErrorResponse
  ErrorResponse func(c *fiber.Ctx, msg string) error
}
```

## Default Config

```go
var ConfigDefault = Config{
  Next:   nil,
  KeyPairs: KeyPairs{},
  AllowMethod: []string{
    fiber.MethodOptions,
    fiber.MethodPost,
    fiber.MethodPut,
    fiber.MethodDelete,
  },
  DebugMode:        false,
  StrictMode:       false,
  ExcludeHealthAPI: true,
  FailResponse:     BadRequestResponse,
  ErrorResponse:    InternalServerErrorResponse,
}
```

## Default Response

```go
func BadRequestResponse(c *fiber.Ctx, msg string) error { // 400
  if msg == "" {
    msg = "Bad Request"
  }
  res := fiber.Map{
    "status":  "bad_request",
    "message": msg,
  }
  return c.Status(fiber.StatusBadRequest).JSON(res)
}

func InternalServerErrorResponse(c *fiber.Ctx, msg string) error { // 500
  if msg == "" {
    msg = "Internal Server Error"
  }
  res := fiber.Map{
    "status":  "internal_server_error",
    "message": msg,
  }
  return c.Status(fiber.StatusInternalServerError).JSON(res)
}
```

## KeyPairs Property

```go
type KeyPairs struct {
  AESKeyForEncrypt []byte
  AESIVForEncrypt  []byte
  AESKeyForDecrypt []byte
  AESIVForDecrypt  []byte
}
```

## Payload Template

An example of payload template (see more how to work in [Example](./example))

### Request

```json
{
  "payload": "FDp1Dl31zGx5nRXFNKihB+k3ly/L7HI9tlHycbKVRwhaf3RRdyFGviuntEZqst0/"
}
```

which can be decrypt to:

```json
{
  "firstname": "Chinnawat",
  "lastname": "Chimdee"
}
```

### Response

```json
{
  "payload": "tpkWPEI6F/nfgUjjtwyKSUf1erxPL6rQt8jG3RitQ1KpvRALfR5YAgQ0CXYkrwLfTid6VdK3SNlffuu/kvI7Hj7br0ur01TUFUWxQ9cl+8U="
}
```

which encrypted from:

```json
{
  "firstname": "Chinnawat [Modified]",
  "lastname": "Chimdee [Modified]"
}
```

Note: This payload using

- AESKeyForEncrypt (used in encrypting response body): `67890123456789012345678901234567`
- AESIVForEncrypt (used in encrypting response body): `6789012345678901`
- AESKeyForDecrypt (used in decrypting request body): `12345678901234567890123456789012`
- AESIVForDecrypt (used in decrypting request body): `1234567890123456`

## Example Usage

Please go to [example/README.md](./example/README.md)


CHANGELOG.md
# Changelog

## v0.4.1

- [Fixed] Bug on missing check strict mode when decrypted

## v0.4.0

- [Added] Configuration `StrictMode` can be set `false` for supporting empty payload and can be set `true` to required the payload `Note: default is false`

## v0.3.0

- [Changed] Configuration variable naming (human error, sry)
- [Removed] Unused directory

## v0.2.0

- [Changed] Using 2-key pair for better securities

## v0.1.0

- [Initial] Encrypted/decrypted a payload on request/response with AES Algorithm


utils_test.go
package cipherPayload

import (
	"bytes"
	"os"
	"testing"

	"github.com/rs/zerolog/log"
	"github.com/stretchr/testify/require"
)

func captureOutput(f func()) string {
	var buf bytes.Buffer
	log.Logger = log.Output(&buf)
	defer func() {
		log.Logger = log.Output(os.Stderr)
	}()
	f()
	return buf.String()
}

type testCaseIsExist struct {
	name     string
	inputArr []string
	inputStr string
	expected bool
}

func TestIsExist(t *testing.T) {
	tests := []testCaseIsExist{
		{
			name:     "Empty Case",
			inputArr: []string{},
			inputStr: "",
			expected: false,
		},
		{
			name:     "Valid Case",
			inputArr: []string{"1", "2", "3"},
			inputStr: "1",
			expected: true,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			inputArr := test.inputArr
			inputStr := test.inputStr
			expected := test.expected

			actual := isExist(inputArr, inputStr)

			require.Equal(t, expected, actual)
		})
	}
}

type testCaseLogger struct {
	name            string
	inputLogType    string
	inputLogMessage string
	expected        string
}

func TestLogger(t *testing.T) {
	tests := []testCaseLogger{
		{
			name:            "Default Case",
			inputLogType:    "",
			inputLogMessage: "",
			expected:        "",
		},
		{
			name:            "Log Error Case",
			inputLogType:    "error",
			inputLogMessage: "test error",
			expected:        "test error",
		},
		{
			name:            "Log Info Case",
			inputLogType:    "info",
			inputLogMessage: "test info",
			expected:        "test info",
		},
		{
			name:            "Log Debug Case",
			inputLogType:    "debug",
			inputLogMessage: "test debug",
			expected:        "test debug",
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			inputLogType := test.inputLogType
			inputLogMessage := test.inputLogMessage
			expected := test.expected

			actual := captureOutput(func() {
				logger := newLogger(true)
				logger.printf(inputLogType, inputLogMessage)
			})

			require.Contains(t, actual, expected)
		})
	}
}


encryption_test.go
package cipherPayload

import (
	"testing"

	"github.com/stretchr/testify/require"
)

var (
	aesKeyForEncrypt []byte
	aesIVForEncrypt  []byte
	aesKeyForDecrypt []byte
	aesIVForDecrypt  []byte

	keys KeyPairs
)

func init() {
	setAValidConfig()
}

func setAValidConfig() {
	aesKeyForEncrypt = []byte("12345678901234567890123456789012")
	aesIVForEncrypt = []byte("1234567890123456")
	aesKeyForDecrypt = []byte("67890123456789012345678901234567")
	aesIVForDecrypt = []byte("6789012345678901")

	keys = KeyPairs{
		aesKeyForEncrypt,
		aesIVForEncrypt,
		aesKeyForDecrypt,
		aesIVForDecrypt,
	}
}

type testCaseEncryption struct {
	name     string
	input    string
	expected string
}

func TestEncryptionAES(t *testing.T) {
	setAValidConfig()

	tests := []testCaseEncryption{
		{
			name:     "Encryption Success",
			input:    "1234567890121",
			expected: "",
		},
		{
			name:     "Encryption Success - Empty Case",
			input:    "",
			expected: "SOMETHING ELSE",
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			input := test.input
			expected := test.expected

			d := NewAESEncryption(keys, false)
			actual, _ := d.Encrypt(input)

			require.NotEqual(t, expected, actual)
		})
	}

	setAValidConfig()
	// Fail Case
	aesKeyPairsBackup := keys
	t.Run("Fail Case", func(t *testing.T) {
		keys = KeyPairs{}
		input := "smtg"

		d := NewAESEncryption(keys)
		_, err := d.Encrypt(input)

		require.Error(t, err)
	})
	keys = aesKeyPairsBackup
}

func TestDecryptionAES(t *testing.T) {
	setAValidConfig()

	tests := []testCaseEncryption{
		{
			name:     "Decryption Success",
			input:    "ldQtpqnEQarFB3RK1JMqSA==",
			expected: "1234567890121",
		},
		{
			name:     "Decryption Success - Empty Case",
			input:    "",
			expected: "",
		},
		{
			name:     "Decryption Fail - Invalid",
			input:    "y/hNE1N17iI",
			expected: "",
		},
		{
			name:     "Decryption Failed",
			input:    "a",
			expected: "",
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			input := test.input
			expected := test.expected

			d := NewAESEncryption(keys)
			actual, _ := d.Decrypt(input)

			require.Equal(t, expected, actual)
		})
	}

	setAValidConfig()
	// Fail Case
	aesKeyPairsBackup := keys
	t.Run("Fail Case", func(t *testing.T) {
		keys = KeyPairs{}
		input := "smtg"

		d := NewAESEncryption(keys)
		_, err := d.Decrypt(input)

		require.Error(t, err)
	})
	keys = aesKeyPairsBackup
}

func TestPkcs7Unpad(t *testing.T) {
	tests := []struct {
		name           string
		inputByte      []byte
		inputBlocksize int
		expected       []byte
		expectedErr    error
	}{
		{
			name:           "Invalid Blocksize - Zero",
			inputByte:      []byte(""),
			inputBlocksize: 0,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidBlockSize,
		},
		{
			name:           "Invalid InputByte - Nil",
			inputByte:      []byte(nil),
			inputBlocksize: 2,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidPKCS7Data,
		},
		{
			name:           "Invalid Padding - Modulo != 0",
			inputByte:      []byte("abc"),
			inputBlocksize: 2,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidPKCS7Padding,
		},
		{
			name:           "Invalid Padding - Trailing with space",
			inputByte:      []byte("abc "),
			inputBlocksize: 4,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidPKCS7Padding,
		},
		{
			name: "Invalid Padding - Trailing with tab space",
			inputByte: []byte("y/hNE1N17iIxSqvM0IWidQ=	"),
			inputBlocksize: 24,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidPKCS7Padding,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			inputByte := test.inputByte
			inputBlocksize := test.inputBlocksize
			expected := test.expected
			expectedErr := test.expectedErr

			d := NewAESEncryption(keys)
			actual, actualErr := d.pkcs7Unpad(inputByte, inputBlocksize)

			require.Equal(t, expected, actual)
			require.Equal(t, expectedErr, actualErr)
		})
	}
}

func TestPkcs7Pad(t *testing.T) {
	tests := []struct {
		name           string
		inputByte      []byte
		inputBlocksize int
		expected       []byte
		expectedErr    error
	}{
		{
			name:           "Invalid Blocksize - Lower than Zero",
			inputByte:      []byte(""),
			inputBlocksize: -1,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidBlockSize,
		},
		{
			name:           "Invalid Data - Empty",
			inputByte:      []byte(nil),
			inputBlocksize: 2,
			expected:       []byte(nil),
			expectedErr:    ErrInvalidPKCS7Data,
		},
	}
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			inputByte := test.inputByte
			inputBlocksize := test.inputBlocksize
			expected := test.expected
			expectedErr := test.expectedErr

			d := NewAESEncryption(keys)
			actual, actualErr := d.pkcs7Pad(inputByte, inputBlocksize)

			require.Equal(t, expected, actual)
			require.Equal(t, expectedErr, actualErr)
		})
	}
}


init.go
package cipherPayload

import (
	"encoding/json"
	"strings"

	"github.com/gofiber/fiber/v2"
)

var (
	serviceName = "[Middleware] cipherPayload"
)

func New(config ...Config) fiber.Handler {
	// set default config
	cfg := configDefault(config...)

	panicResponseHeader := serviceName + ": Some configuration is missing: "

	// config is required
	if len(cfg.KeyPairs.AESKeyForEncrypt) == 0 {
		panic(panicResponseHeader + "`AESKeyForEncrypt` is required.")
	}

	// config is required
	if len(cfg.KeyPairs.AESIVForEncrypt) == 0 {
		panic(panicResponseHeader + "`AESIVForEncrypt` is required.")
	}

	// config is required
	if len(cfg.KeyPairs.AESKeyForDecrypt) == 0 {
		panic(panicResponseHeader + "`AESKeyForDecrypt` is required.")
	}

	// config is required
	if len(cfg.KeyPairs.AESIVForDecrypt) == 0 {
		panic(panicResponseHeader + "`AESIVForDecrypt` is required.")
	}

	return func(c *fiber.Ctx) error {
		// Don't execute middleware if Next returns true
		if cfg.Next != nil && cfg.Next(c) {
			return c.Next()
		}

		// Don't execute when url is contains "health"
		if strings.Contains(c.OriginalURL(), "health") {
			return c.Next()
		}

		// Don't execute when the method is not matched a list
		if !isExist(cfg.AllowMethod, c.Method()) {
			return c.Next()
		}

		logger := newLogger(cfg.DebugMode)

		var reqBody PayloadBody
		err := c.BodyParser(&reqBody)
		if err != nil || reqBody.Payload == "" {
			errMsg := "BodyParser error or Empty Payload"
			if cfg.StrictMode {
				errMsg := "Invalid Payload"
				logger.printf("error", errMsg, string(c.Request().Body()))
				return cfg.FailResponse(c, errMsg)
			}
			logger.printf("info", errMsg, string(c.Request().Body()))
		}
		logger.printf("debug", "Request:", reqBody.Payload)

		// Payload Decrypting
		encrypterDecrypter := NewAESEncryption(cfg.KeyPairs, cfg.DebugMode)
		decryptedPayload, err := encrypterDecrypter.Decrypt(reqBody.Payload)
		logger.printf("debug", "Decrypted:", decryptedPayload)
		if err != nil || decryptedPayload == "" {
			errMsg := "Payload decrypt fail or Empty payload"
			if cfg.StrictMode {
				logger.printf("error", serviceName, err)
				return cfg.FailResponse(c, errMsg)
			}
			logger.printf("info", errMsg)
		}

		jsonRaw := json.RawMessage(decryptedPayload)
		jsonBytes, _ := json.Marshal(jsonRaw)

		// Set plaintext back into request body
		logger.printf("debug", "jsonBytes:", string(jsonBytes))
		c.Request().SetBodyRaw(jsonBytes)

		// Let request to continue execute
		c.Next()

		// Intercept the response body
		interceptBody := string(c.Response().Body())
		logger.printf("debug", "Intercept Body:", interceptBody)

		// Payload Encrypting
		encryptedPayload, err := encrypterDecrypter.Encrypt(interceptBody)
		logger.printf("debug", "Encrypted:", encryptedPayload)

		if err != nil || encryptedPayload == "" {
			logger.printf("error", serviceName, err)
			errMsg := "InternalServerError: Cannot encrypt payload or invalid payload"
			return cfg.ErrorResponse(c, errMsg)
		}

		// Set ciphertext back into response body
		var resBody PayloadBody
		resBody.Payload = encryptedPayload
		return c.JSON(resBody)
	}
}


utils.go
package cipherPayload

import (
	"fmt"
	"strings"

	"github.com/rs/zerolog/log"
)

func isExist(slice []string, target string) bool {
	for _, item := range slice {
		if item == target {
			return true
		}
	}
	return false
}

func recoveryCatch() {
	if r := recover(); r != nil {
		log.Debug().Msgf("Recovered in f: %v", r)
	}
}

// Logger
type logger interface {
	printf(logType string, anything ...interface{})
}

type loggerConfig struct {
	isShow bool
}

func newLogger(isShow bool) logger {
	return &loggerConfig{
		isShow: isShow,
	}
}

func (l *loggerConfig) printf(logType string, anything ...interface{}) {
	if !l.isShow {
		return
	}

	var logList []string
	for _, val := range anything {
		valStr := fmt.Sprintf("%v", val)
		logList = append(logList, valStr)
	}

	logHeader := "[CipherPayload]"
	logString := strings.Join(logList, " ")
	switch logType {
	case "error":
		log.Error().Msgf("%v %v", logHeader, logString)
	case "info":
		log.Info().Msgf("%v %v", logHeader, logString)
	default:
		log.Debug().Msgf("%v %v", logHeader, logString)
	}
}


encryption.go
package cipherPayload

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"encoding/base64"
	"errors"
)

var (
	// ErrInvalidBlockSize indicates hash blocksize <= 0.
	ErrInvalidBlockSize = errors.New("invalid blocksize")

	// ErrInvalidPKCS7Data indicates bad input to PKCS7 pad or unpad.
	ErrInvalidPKCS7Data = errors.New("invalid PKCS7 data (empty or not padded)")

	// ErrInvalidPKCS7Padding indicates PKCS7 unpad fails to bad input.
	ErrInvalidPKCS7Padding = errors.New("invalid padding on input")
)

type AES interface {
	Encrypt(plainText string) (string, error)
	Decrypt(encryptedText string) (string, error)
	pkcs7Unpad(b []byte, blocksize int) ([]byte, error)
	pkcs7Pad(b []byte, blocksize int) ([]byte, error)
}

type DefaultAESEncryption struct {
	keys   KeyPairs
	logger logger
}

func NewAESEncryption(keys KeyPairs, debugMode ...bool) AES {
	isShowLog := false
	if len(debugMode) > 0 {
		isShowLog = bool(debugMode[0])
	}
	logger := newLogger(isShowLog)
	return &DefaultAESEncryption{
		keys:   keys,
		logger: logger,
	}
}

func (d *DefaultAESEncryption) Encrypt(plainText string) (string, error) {
	serviceName := "Encryption - AES - Encrypt"
	d.logger.printf("info", serviceName)

	if plainText == "" {
		return "", nil
	}

	byteText := []byte(plainText)
	block, err := aes.NewCipher(d.keys.AESKeyForEncrypt)
	if err != nil {
		d.logger.printf("error", serviceName, err)
		return "", err
	}

	defer recoveryCatch()
	byteText, _ = d.pkcs7Pad(byteText, block.BlockSize())

	ciphertext := make([]byte, len(byteText))

	bm := cipher.NewCBCEncrypter(block, d.keys.AESIVForEncrypt)
	bm.CryptBlocks(ciphertext, byteText)

	cipherText := base64.StdEncoding.EncodeToString(ciphertext)
	d.logger.printf("debug", "Encrypted:", cipherText)
	return cipherText, nil
}

func (d *DefaultAESEncryption) Decrypt(encryptedText string) (string, error) {
	serviceName := "Encryption - AES - Decrypt"
	d.logger.printf("info", serviceName)

	if encryptedText == "" {
		return "", nil
	}
	ciphertext, _ := base64.StdEncoding.DecodeString(encryptedText)
	block, err := aes.NewCipher(d.keys.AESKeyForDecrypt)
	if err != nil {
		d.logger.printf("error", serviceName, err)
		return "", err
	}

	defer recoveryCatch()
	bm := cipher.NewCBCDecrypter(block, d.keys.AESIVForDecrypt)
	bm.CryptBlocks(ciphertext, ciphertext)

	out, err := d.pkcs7Unpad(ciphertext, aes.BlockSize)
	if err != nil {
		d.logger.printf("error", serviceName, err)
		return "", err
	}

	plainText := string(out)
	d.logger.printf("debug", "Decrypted:", plainText)
	return plainText, nil
}

func (d DefaultAESEncryption) pkcs7Unpad(b []byte, blocksize int) ([]byte, error) {
	if blocksize <= 0 {
		return nil, ErrInvalidBlockSize
	}
	if b == nil {
		return nil, ErrInvalidPKCS7Data
	}
	if len(b) == 0 {
		return nil, ErrInvalidPKCS7Data
	}
	if len(b)%blocksize != 0 {
		return nil, ErrInvalidPKCS7Padding
	}
	c := b[len(b)-1] // the last character
	n := int(c)      // convert the last char into a number
	if n == 0 || n > len(b) {
		return nil, ErrInvalidPKCS7Padding
	}
	for i := 0; i < n; i++ {
		if b[len(b)-n+i] != c {
			return nil, ErrInvalidPKCS7Padding
		}
	}
	return b[:len(b)-n], nil
}

func (d DefaultAESEncryption) pkcs7Pad(b []byte, blocksize int) ([]byte, error) {
	if blocksize <= 0 {
		return nil, ErrInvalidBlockSize
	}
	if b == nil {
		return nil, ErrInvalidPKCS7Data
	}
	n := blocksize - (len(b) % blocksize)
	pb := make([]byte, len(b)+n)
	copy(pb, b)
	copy(pb[len(b):], bytes.Repeat([]byte{byte(n)}, n))
	return pb, nil
}


example/main.go
package main

import (
	"fmt"

	"github.com/buildingwatsize/cipherPayload"
	"github.com/gofiber/fiber/v2"
	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
)

var (
	aesSecretKeyForEncrypt []byte
	aesSecretIVForEncrypt  []byte
	aesSecretKeyForDecrypt []byte
	aesSecretIVForDecrypt  []byte
)

func init() {
	aesSecretKeyForEncrypt = []byte("12345678901234567890123456789012")
	aesSecretIVForEncrypt = []byte("1234567890123456")
	aesSecretKeyForDecrypt = []byte("67890123456789012345678901234567")
	aesSecretIVForDecrypt = []byte("6789012345678901")

	// UNIX Time is faster and smaller than most timestamps
	zerolog.TimeFieldFormat = zerolog.TimeFormatUnix
}

func main() {
	// NOTE: For more usage please go to README.md

	// Without cipherPayload middleware
	// ================================
	appPlaintext := fiber.New()
	apiPlaintextGroup := appPlaintext.Group("/api", nextHandlerAPI) // /api
	apiPlaintextGroup.Post("/example", HandlerExamplePlaintext)     // /api/example (FullPath = http://localhost:5000/api/example)

	// With cipherPayload middleware
	// =============================
	appCipher := fiber.New()
	appCipher.Use(cipherPayload.New(cipherPayload.Config{
		KeyPairs: cipherPayload.KeyPairs{
			AESKeyForEncrypt: aesSecretKeyForEncrypt,
			AESIVForEncrypt:  aesSecretIVForEncrypt,
			AESKeyForDecrypt: aesSecretKeyForDecrypt,
			AESIVForDecrypt:  aesSecretIVForDecrypt,
		},
		DebugMode: true,
	}))
	apiCipherGroup := appCipher.Group("/api", nextHandlerAPI) // /api
	apiCipherGroup.Post("/example", HandlerExampleCipher)     // /api/example (FullPath = http://localhost:8000/api/example)

	PortPlaintext := "5000"
	go appPlaintext.Listen(":" + PortPlaintext)
	PortCipher := "8000"
	appCipher.Listen(":" + PortCipher)
}

var nextHandlerAPI = func(c *fiber.Ctx) error {
	c.Accepts(fiber.MIMEApplicationJSONCharsetUTF8)

	log.Info().Msg("Called API")
	return c.Next()
}

var HandlerExamplePlaintext = func(c *fiber.Ctx) error {
	log.Info().Msg("Called HandlerExamplePlaintext")

	var reqBody map[string]interface{}
	_ = c.BodyParser(&reqBody)
	for k, v := range reqBody {
		reqBody[k] = fmt.Sprintf("%v [Modified]", v)
	}

	return c.Status(fiber.StatusOK).JSON(reqBody)
}

// ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
// === YOU SEE ? HOW IT DIFFERENCE ? ===
// :: PS. the "Plaintext Way" and the "Cipher Way"
//        can be used the same handler ::
// vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv

var HandlerExampleCipher = func(c *fiber.Ctx) error {
	log.Info().Msg("Called HandlerExampleCipher")

	var reqBody map[string]interface{}
	_ = c.BodyParser(&reqBody)
	for k, v := range reqBody {
		reqBody[k] = fmt.Sprintf("%v [Modified]", v)
	}

	return c.Status(fiber.StatusOK).JSON(reqBody)
}


config.go
package cipherPayload

import (
	"github.com/gofiber/fiber/v2"
)

type Config struct {
	// Optional. Default: nil
	Next func(c *fiber.Ctx) bool

	// Required. Default: KeyPairs{}
	KeyPairs KeyPairs

	// Optional. Default: OPTIONS, POST, PUT, DELETE
	AllowMethod []string

	// Optional. Default: false
	DebugMode bool

	// Optional. [Default: false]
	StrictMode bool

	// Optional. Default: true
	ExcludeHealthAPI bool

	// Optional. Default: BadRequestResponse
	FailResponse func(c *fiber.Ctx, msg string) error

	// Optional. Default: InternalServerErrorResponse
	ErrorResponse func(c *fiber.Ctx, msg string) error
}

var ConfigDefault = Config{
	Next:     nil,
	KeyPairs: KeyPairs{},
	AllowMethod: []string{
		fiber.MethodOptions,
		fiber.MethodPost,
		fiber.MethodPut,
		fiber.MethodDelete,
	},
	DebugMode:        false,
	StrictMode:       false,
	ExcludeHealthAPI: true,
	FailResponse:     BadRequestResponse,
	ErrorResponse:    InternalServerErrorResponse,
}

func configDefault(config ...Config) Config {
	// Return default config if nothing provided
	if len(config) < 1 {
		return ConfigDefault
	}

	// Override default config
	cfg := config[0]

	// set default values
	if cfg.AllowMethod == nil {
		cfg.AllowMethod = ConfigDefault.AllowMethod
	}

	// Note: cfg.DebugMode: it's false by default.
	// Note: cfg.StrictMode: it's false by default.
	// NOTE: cfg.ExcludeHealthAPI: it's true by default

	// set default values
	if cfg.FailResponse == nil {
		cfg.FailResponse = ConfigDefault.FailResponse
	}

	// set default values
	if cfg.ErrorResponse == nil {
		cfg.ErrorResponse = ConfigDefault.ErrorResponse
	}

	return cfg
}

type PayloadBody struct {
	Payload string `json:"payload"`
}

type KeyPairs struct {
	AESKeyForEncrypt []byte
	AESIVForEncrypt  []byte
	AESKeyForDecrypt []byte
	AESIVForDecrypt  []byte
}

func BadRequestResponse(c *fiber.Ctx, msg string) error { // 400
	if msg == "" {
		msg = "Bad Request"
	}
	res := fiber.Map{
		"status":  "bad_request",
		"message": msg,
	}
	return c.Status(fiber.StatusBadRequest).JSON(res)
}

func InternalServerErrorResponse(c *fiber.Ctx, msg string) error { // 500
	if msg == "" {
		msg = "Internal Server Error"
	}
	res := fiber.Map{
		"status":  "internal_server_error",
		"message": msg,
	}
	return c.Status(fiber.StatusInternalServerError).JSON(res)
}
